---
layout: docs
title: References
prev: part8
number: 9
---

# References

Bailey TL, Boden M, Buske FA, Frith M, Grant CE, Clementi L, Ren J, Li WW, Noble WS (2009). “MEME SUITE: tools for motif discovery and searching.” *Nucleic acids research*, p. gkp335.

Bailey TL, Williams N, Misleh C, Li WW (2006). “MEME: discovering and analyzing DNA and protein sequence motifs.” *Nucleic acids research*, **34**(suppl 2), W369–W373.

Buenrostro JD, Giresi PG, Zaba LC, Chang HY, Greenleaf WJ (2013). “Transposition of native chromatin for fast and sensitive epigenomic profiling of open chromatin, DNA-binding proteins and nucleosome position.” *Nature methods*, **10**(12), 1213–1218.

Core LJ, Martins AL, Danko CG, Waters CT, Siepel A, Lis JT (2014). “Analysis of nascent RNA identifies a unified architecture of initiation regions at mammalian promoters and enhancers.” *Nature genetics*, **46**(12), 1311–1320.

Core LJ, Waterfall JJ, Lis JT (2008). “Nascent RNA sequencing reveals widespread pausing and divergent initiation at human promoters.” *Science*, **322**(5909), 1845–1848.

Grant CE, Bailey TL, Noble WS (2011). “FIMO: scanning for occurrences of a given motif.” *Bioinformatics*, **27**(7), 1017–1018.

Grøntved L, Bandle R, John S, Baek S, Chung HJ, Liu Y, Aguilera G, Oberholtzer C, Hager GL, Levens D (2012). “Rapid genome-scale mapping of chromatin accessibility in tissue.” *Epigenetics & chromatin*, **5**(1), 1.

Grøntved L, John S, Baek S, Liu Y, Buckley JR, Vinson C, Aguilera G, Hager GL (2013). “C/EBP maintains chromatin accessibility in liver and facilitates glucocorticoid receptor recruitment to steroid response elements.” *The EMBO journal*, **32**(11), 1568–1583.

He HH, Meyer CA, Chen MW, Zang C, Liu Y, Rao PK, Fei T, Xu H, Long H, Liu XS, et al. (2014). “Refined DNase-seq protocol and data analysis reveals intrinsic bias in transcription factor footprint identification.” *Nature methods*, **11**(1), 73–78.

Karolchik D, Barber GP, Casper J, Clawson H, Cline MS, Diekhans M, Dreszer TR, Fujita PA, Guruvadoo L, Haeussler M, et al. (2014). “The UCSC genome browser database: 2014 update.” *Nucleic acids research*, **42**(D1), D764–D770.

Kwak H, Fuda NJ, Core LJ, Lis JT (2013). “Precise maps of RNA polymerase reveal how promoters direct initiation and pausing.” *Science*, **339**(6122), 950–953.

Langmead B, Trapnell C, Pop M, Salzberg S (2009). “Ultrafast and memory-efficient alignment of short DNA sequences to the human genome.” *Genome Biology*, **10**(3), R25. ISSN 1465-6906. doi: 10.1186/gb-2009-10-3-r25. URL http://genomebiology.com/2009/10/3/R25.

Lazarovici A, Zhou T, Shafer A, Machado ACD, Riley TR, Sandstrom R, Sabo PJ, Lu Y, Rohs R, Stamatoyannopoulos JA, et al. (2013). “Probing DNA shape and methylation state on a genomic scale with DNase I.” *Proceedings of the National Academy of Sciences*, **110**(16), 6376–6381.

Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, Marth G, Abecasis G, Durbin R, et al. (2009). “The sequence alignment/map format and SAMtools.” *Bioinformatics*, **25**(16), 2078–2079.

Neph S, Vierstra J, Stergachis AB, Reynolds AP, Haugen E, Vernot B, Thurman RE, John S, Sandstrom R, Johnson AK, Maurano MT, Humbert R, Rynes E, Wang H, Vong S, Lee K, Bates D, Diegel M, Roach V, Dunn D, Neri J, Schafer A, Hansen RS, Kutyavin T, Giste E, Weaver M, Canfield T, Sabo P, Zhang M, Balasundaram G, Byron R, MacCoss MJ, Akey JM, Bender MA, Groudine M, Kaul R, Stamatoyannopoulos JA (2012). “An expansive human regulatory lexicon encoded in transcription factor footprints.” *Nature*, **489**(7414), 83–90.

Quinlan AR, Hall IM (2010). “BEDTools: a flexible suite of utilities for comparing genomic features.” *Bioinformatics*, **26**(6), 841–842.

Seo YK, Chong HK, Infante AM, Im SS, Xie X, Osborne TF (2009). “Genome-wide analysis of SREBP1 binding in mouse liver chromatin reveals a preference for promoter proximal binding to a new motif.” *Proceedings of the National Academy of Sciences*, **106**(33), 13765–13769.

Stamatoyannopoulos JA, Snyder M, Hardison R, Ren B, Gingeras T, Gilbert DM, Groudine M, Bender M, Kaul R, Canfield T, et al. (2012). “An encyclopedia of mouse DNA elements (Mouse ENCODE).” *Genome biology*, **13**(8), 1.

Sung MH, Guertin MJ, Baek S, Hager GL (2014). “DNase footprint signatures are dictated by factor dynamics and DNA sequence.” *Molecular cell*, **56**(2), 275–285.

Thomas CA (1956). “The Enzymatic Degradation of Desoxyribose Nucleic Acid.” *J. Am. Chem. Soc.*, **78**(9), 1861–1868.

Vierstra J, Stamatoyannopoulos JA (2016). “Genomic footprinting.” *Nat. Methods*, **13**(3), 213–221.

Yardımcı GG, Frank CL, Crawford GE, Ohler U (2014). “Explicit DNase sequence bias modeling enables high-resolution transcription factor footprint detection.” *Nucleic acids research*, p. gku810.
